### ISSUE TITLE
[BUG] 使用本repo提供的finetune.py脚本训练的模型，推理过程的头几个token重复输出<im_start>

### ISSUE THREAD

**Issue by ericzhou571** [QwenLM/Qwen#446]
### 是否已有关于该错误的issue或讨论？ | Is there an existing issue / discussion for this?

- [X] 我已经搜索过已有的issues和讨论 | I have searched the existing issues / discussions

### 该问题是否在FAQ中有解答？ | Is there an existing answer for this in FAQ?

- [X] 我已经搜索过FAQ | I have searched FAQ

### 当前行为 | Current Behavior

训练后的模型输出结果示例：
```
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
<|im_start|>
result:
{"task_name":"抬头显示位置.....
```
在重复输出了多次<im_start>后才开始正常输出completion。

我们已经确认训练数据的completion并未包含任何<im_start>字符。

通过检查finetune.py
https://github.com/QwenLM/Qwen/blob/dbc111267a6018dc2eda1fbf79aa59b859c72640/finetune.py#L149-L160
以一个训练数据为例（mask token 使用#进行替代），使用上述preprocess中的mask方法，mask后的target结果如下：

![截屏2023-10-11 15 55 13](https://github.com/QwenLM/Qwen/assets/57415741/17878958-ac1a-4a56-be02-d115289b8bd5)
我们发现在sft训练数据的target 构建过程中，不论是user或者assistant的拼接过程中，<|im_start|> 都作为预测目标出现，但是之后的角色assisstant又仍然被mask掉了，请问这样的target设定是大佬有意为之吗？

参考qwen模型的chat代码qwen_generation_utils.py的 make_context函数，<im_start>都是在推理拼接prompt的时候就拼接到了prompt里，似乎不需要模型来预测出<im_start>这个token，不知道是否是大佬们忘记mask掉<im_start>了？
```
        context_tokens += (
            nl_tokens
            + im_start_tokens
            + _tokenize_str("user", query)[1]
            + im_end_tokens
            + nl_tokens
            + im_start_tokens
            + tokenizer.encode("assistant")
            + nl_tokens
        )
        raw_text += f"\n{im_start}user\n{query}{im_end}\n{im_start}assistant\n"
```

### 期望行为 | Expected Behavior
@JustinLin610 
期望大佬们确认一下preprocess的mask方案是否合理，是否能解决模型前几个token为<im_start>的问题

### 复现方法 | Steps To Reproduce

1. 使用finetune.py 训练模型
2. 使用temperature 0 推理finetune后的模型

### 运行环境 | Environment

```Markdown
- OS: Ubuntu 20.04
- Python: 3.8
- Transformers: 4.31.0
- PyTorch: 2.0.1
- CUDA (`python -c 'import torch; print(torch.version.cuda)'`): 11。7
```

### 备注 | Anything else?

无

**Comment by jklj077** [comment]
temperature=0 这个使用场景，目前还不支持哈。请问调到0.5、0.8这种的话，还会有这个现象吗？

**Comment by ericzhou571** [comment]
> temperature=0 这个使用场景，目前还不支持哈。请问调到0.5、0.8这种的话，还会有这个现象吗？

感谢大佬快速回复，之前将temperature调节到0.1 和 0.3 仍然会持续输出<im_start>， 

0.5 0.8 晚些时候测一下。

但是这个问题是偶发的，600多个推理样本里，100多个会出现。

**所以大佬这边确认下来是preprocess的mask方案本身没有问题吗？**

**Comment by jklj077** [comment]
> preprocess的mask方案本身没有问题吗？

开源的Chat模型也是这个策略，还没遇到有这样的问题。

finetune方式是什么呢？全量微调、lora、qlora，从预训练模型还是chat模型开始训练的？

**Comment by jklj077** [comment]
先关掉了，如问题仍存在，请重开本issue吧。
